  ---
title: "Analysis of Hypoxia and Normoxia on Microarray cell lines (E-GEOD-18494)"
output:
  pdf_document: default
  html_document: 
    default
  github_document: 
    df_print: paged
    html_preview: FALSE
    keep_html: TRUE
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_format = "all") })    
---
      

Expression profiling of hypoxic HepG2 hepatoma, U87 glioma, and MDA-MB231 breast cancer cells: time course (E-GEOD-18494)

Analysis of expression changes of cultured HepG2 hepatoma, U87 glioma, and MDA-MB231 breast cancer cells subjected to hypoxia (0.5% O2) for 0, 4, 8, 12 hours . Results provide insight to cell type-specific response to hypoxia. HepG2 hepatoma, U87 glioma, and MDA-MB231 breast cancer cells were collected under normoxic conditions (~19% O2, 0 hours) and after 4, 8 and 12 hours of hypoxia treatment (0.5% O2). For each cell line, three replicates of total RNA at each time point were prepared using Trizol and submitted to the DFCI Microarray Core for labeling, hybridization to Affymetrix HG-U133Plus2 oligonucleotide arrays and image scanning.
    
https://www.ebi.ac.uk/arrayexpress/experiments/E-GEOD-18494/   


This pipeline if based on the Workflow : 

https://www.bioconductor.org/packages/devel/workflows/vignettes/maEndToEnd/inst/doc/MA-Workflow.html

  
  ```{r error=TRUE, message=FALSE, warning=FALSE, include=FALSE, purl=FALSE, results='hide'}
  ## This chunk automatically generates a text .R version of this script when running within knitr.
  input  = knitr::current_input()  # filename of input document
  output = paste(tools::file_path_sans_ext(input), 'R', sep = '.')
  try(knitr::purl(input,output,documentation=2,quiet=T), silent = TRUE)
  # Avoid duplicate label error of knitr::purl
  options(knitr.duplicate.label = 'allow')
  # Code to browse the markdown file with renderized images.
  knitr::opts_chunk$set(
    fig.path = "figs/render-"
  )
  ```

    
```{r message=FALSE, warning=FALSE}

packages_bioconductor = c("Biobase","ArrayExpress", "hgu133plus2.db", "oligo", "DESeq2", "Glimma", "limma", "clusterProfiler", "topGO")

# Install and load packages
package.check <- lapply(packages_bioconductor, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    BiocManager::install(x, dependencies = TRUE)
    library(x, character.only = TRUE)
  }
})

packages_cran = c("DT", "tidyverse","dplyr", "tibble", "data.table", "stats", "ggpubr", "ggplot2", "RColorBrewer", "ggrepel", "pheatmap","ComplexHeatmap", "gapminder", "ggstatsplot", "fs")

# For ggstatsplot, might be needed execute:
# sudo apt-get install libgmp-dev
# sudo apt-get install libmpfr-dev

# Install and load packages
package.check <- lapply(packages_cran, FUN = function(x) {
  if (!require(x, character.only = TRUE)) {
    install.packages(x, dependencies = TRUE)
    library(x, character.only = TRUE)
  }
})

rm(package.check, packages_bioconductor, packages_cran)

```
    
    
# Import of annotation data and microarray expression data as "ExpressionSet"


```{r message=FALSE, warning=FALSE }
download_dir <- fs::path(".data_tmp")
if (!dir_exists(download_dir)) {
    dir_create(download_dir)
    EGEOD18494 <- ArrayExpress( "E-GEOD-18494", save=TRUE, path=download_dir)
} else {
    EGEOD18494 <- ArrayExpress( "E-GEOD-18494", save=TRUE, path=download_dir)
}

data.EGEOD18494 <- pData(EGEOD18494)

data.EGEOD18494 <- data.frame(
                  codes = substr(data.EGEOD18494$Source.Name,1,9),
                  cell_line = data.EGEOD18494$Characteristics..cell.line.,
                  time = data.EGEOD18494$Characteristics..time,
                  condition = data.EGEOD18494$Characteristics..stress.
                  )
data.EGEOD18494 <- data.EGEOD18494[order(data.EGEOD18494$codes),]
data.EGEOD18494$rep <- rep(1:3, n= length(data.EGEOD18494$codes))


eset.EGEOD18494 <- oligo::rma(EGEOD18494,  normalize = FALSE)

exp.EGEOD18494 <- exprs(eset.EGEOD18494)

colnames(exp.EGEOD18494) <- substr(colnames(exp.EGEOD18494),1,9)

EGEOD18494@annotation

rm(download_dir)
```



```{r}
row_medians_assayData <- 
  Biobase::rowMedians(as.matrix(Biobase::exprs(eset.EGEOD18494)))

RLE_data <- sweep(Biobase::exprs(eset.EGEOD18494), 1, row_medians_assayData)

RLE_data <- as.data.frame(RLE_data)
RLE_data_gathered <- 
  tidyr::gather(RLE_data, patient_array, log2_expression_deviation)

ggplot2::ggplot(RLE_data_gathered, aes(patient_array,
                                       log2_expression_deviation)) + 
  geom_boxplot(outlier.shape = NA) + 
  ylim(c(-2, 2)) + 
  theme(axis.text.x = element_text(colour = "aquamarine4", 
                                  angle = 60, size = 6.5, hjust = 1 ,
                                  face = "bold"))

rm(RLE_data, RLE_data_gathered, row_medians_assayData)
```

Note that the y-axis now displays for each sample the deviation of expression intensity from the median expression of the respective single transcripts across arrays.

Boxes with a larger extension therefore indicate an unusually high deviation from the median in a lot of transcripts, suggesting that these arrays are different from most of the others in some way.

Boxes that are shifted in y-direction indicate a systematically higher or lower expression of the majority of transcripts in comparison to most of the other arrays. This could be caused by quality issues or batch effects.

Therefore, if shape and median of a given box varies too much from the bulk, they should be inspected and potentially removed.

Checking the intensity distributions of the individual arrays, it they are quite different, this indicates the need for an appropriate normalization.


```{r}
EGEOD18494_norm <- oligo::rma(EGEOD18494, normalize = TRUE)

rm(eset.EGEOD18494)
```


```{r}
row_medians_assayData <- 
  Biobase::rowMedians(as.matrix(Biobase::exprs(EGEOD18494_norm)))

RLE_data <- sweep(Biobase::exprs(EGEOD18494_norm), 1, row_medians_assayData)

RLE_data <- as.data.frame(RLE_data)
RLE_data_gathered <- 
  tidyr::gather(RLE_data, patient_array, log2_expression_deviation)

ggplot2::ggplot(RLE_data_gathered, aes(patient_array,
                                       log2_expression_deviation)) + 
  geom_boxplot(outlier.shape = NA) + 
  ylim(c(-2, 2)) + 
  theme(axis.text.x = element_text(colour = "aquamarine4", 
                                  angle = 60, size = 6.5, hjust = 1 ,
                                  face = "bold"))
rm(RLE_data, RLE_data_gathered, row_medians_assayData)
```


```{r}

anno.EGEOD18494 <- AnnotationDbi::select(hgu133plus2.db, keys=rownames(exp.EGEOD18494), columns=c("ENSEMBL", "SYMBOL", "GENENAME"), keytype="PROBEID")

colnames(anno.EGEOD18494) <- c("probes", "ensgene", "symbol", "description")

```

# Quality control of the raw data: PCA of Samples
    
```{r,  fig.height = 10, fig.width=15}

PCA_raw <- prcomp(t(exp.EGEOD18494), scale. = FALSE)

percentVar <- round(100*PCA_raw$sdev^2/sum(PCA_raw$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- cbind(data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2]), data.EGEOD18494)

col3 <- guide_legend(ncol = 3)
ggplot(dataGG, aes(PC1, PC2)) +
  geom_point(aes(shape = cell_line, colour = time, size = condition)) +
# geom_point(colour = condition, size = 1) +
  ggtitle("PCA plot of the log-transformed raw expression data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5))+
  coord_fixed(ratio = sd_ratio) +
  scale_shape_manual(values = c(0, 1, 2), guide = col3) + 
  scale_color_manual(values = c("darkorange2", "dodgerblue4", "chartreuse4", "chartreuse1"), guide = col3) +
  guides(fill=guide_legend(ncol=3)) +
  theme(
    legend.title = element_text(size = 10),
    legend.text = element_text(size = 8)
  ) 

rm(col3, dataGG, PCA_raw, sd_ratio, percentVar)
```
    
# MultiDimension Scaling plots by conditions and time
    
```{r,  fig.height = 10, fig.width=15}

dist_matrix<- dist(t(exp.EGEOD18494)) # very important to take transpose "t"
mds <- cmdscale(dist_matrix)

plot(mds[,1],mds[,2],bg=as.numeric(data.EGEOD18494$condition),
     pch=21,xlab="First dimension",ylab="Second dimension",
     cex=2,main="MultiDimension Scaling PLot: Hypoxia vs Normoxia")
legend("topright",levels(data.EGEOD18494$condition),col=seq(along=levels(data.EGEOD18494$condition)),pch=15,cex=1)
identify(mds)

plot(mds[,1],mds[,2],bg=as.numeric(data.EGEOD18494$time),
     pch=21,xlab="First dimension",ylab="Second dimension",
     cex=2,main="MultiDimension Scaling PLot: Hypoxia vs Normoxia")
legend("topright",levels(data.EGEOD18494$time),col=seq(along=levels(data.EGEOD18494$time)),pch=15,cex=1)
identify(mds)

rm(mds,dist_matrix)
```

# Heatmap clustering analysis
    
```{r,  fig.height = 10, fig.width=15}

annotation_for_heatmap <- 
  data.frame(condition = data.EGEOD18494$condition,  
             time = data.EGEOD18494$time, 
             cell_line = data.EGEOD18494$cell_line
             )

row.names(annotation_for_heatmap) <- colnames(exp.EGEOD18494)

dists <- as.matrix(dist(t(exp.EGEOD18494), method = "manhattan"))

rownames(dists) <- colnames(exp.EGEOD18494)
hmcol <- rev(colorRampPalette(RColorBrewer::brewer.pal(9, "YlOrRd"))(255))
colnames(dists) <- NULL
diag(dists) <- NA

ann_colors <- list(  
  condition = brewer.pal(length(levels(data.EGEOD18494$condition)), "Dark2"),
  time = brewer.pal(length(levels(data.EGEOD18494$time)), "Set1"),
  cell_line = brewer.pal(length(levels(data.EGEOD18494$cell_line)), "Set2")
)

names(ann_colors$condition) <- levels(data.EGEOD18494$condition)    
names(ann_colors$time) <- levels(data.EGEOD18494$time)
names(ann_colors$cell_line) <- levels(data.EGEOD18494$cell_line)

pheatmap(dists, col = (hmcol), 
         annotation_row = annotation_for_heatmap,
         annotation_colors = ann_colors,
         legend = TRUE, 
         treeheight_row = 0,
         legend_breaks = c(min(dists, na.rm = TRUE), 
                           max(dists, na.rm = TRUE)), 
         legend_labels = (c("small distance", "large distance")),
         main = "Clustering of Cell lines",
         fontsize_row = 5)

rm(ann_colors, hmcol, dists, annotation_for_heatmap )
    
```
    

# Selecting the HIF Genes 
    
```{r}
# Genes from "Dynamical analysis of a generic Boolean model for the control of the mammalian cell cycle"
# Adrien Fauré  1 , Aurélien Naldi, Claudine Chaouiya, Denis Thieffry
# PMID: 16873462 DOI: 10.1093/bioinformatics/btl210 
# CycD Rb E2F CycE CycA p27 Cdc20 Cdh1 UbcH10 CycB

# Genes from Our Boolean Network:    
# HIF1a, HIF2a, p53, BNIP3, VEGF, cMyc, Oct4, cdc20, cycA, cycB, cycE, cycD, p27, Rb, E2F, cdh1, mdm2, BAD, BclX

hif.symbols <- c("HIF1A", "HIF1", "PASD8", "MOP1", "EPAS1", "HIF2A", "HLF", "PASD2", "MOP2", "VEGFA", "VPF", "MVCD1", "VEGF-A", "TP53", "P53", "MYC", "C-Myc", "POU5F1", "OCT3", "OTF3", "CDC20", "P55CDC", "CDC20A", "CCNA1", "CCNA1", "CCNA2", "CCN1", "CCNA", "CCNB1", "CCNB", "CCNB1", "CCNB2", "CCNB2", "HsT17299", "CCND1", "PRAD1", "CCND2", "MPPH3", "CCNE1", "CCNE1", "CCNE", "PCCNE1", "CCNE2", "CCNE2", "CCNE", "PCCNE1", "CDKN1B", "KIP1", "P27KIP1", "CDKN4MEN4", "RB1", "PPP1R130", "Pp110", "E2F1", "RBAP-1", "BNIP3", "NIP3", "BCL2", "MCL1", "BCL2L3", "CDH1", "CD324", "UBE2C", "BAD", "BCL2L8", "VHL", "PVHL", "VHL1", "MDM2", "HDM2", "EP300", "P300", "KAT3B")

hif.probes <- anno.EGEOD18494$probes[anno.EGEOD18494$symbol %in% hif.symbols]

exp.EGEOD18494.hif <- data.frame(exp.EGEOD18494) %>% 
  rownames_to_column('probes') %>% 
  filter(probes %in% hif.probes) %>% 
  merge(anno.EGEOD18494[anno.EGEOD18494$symbol %in% hif.symbols, c("probes","symbol")], by = "probes") %>% 
  distinct(symbol, .keep_all = TRUE) %>%
  select(!(probes)) 

# Saving data with data description on columns
exp.EGEOD18494.hif %>%  
  rename_at(vars(data.EGEOD18494$codes), 
            ~paste0(substr(data.EGEOD18494$condition,1,4),".",
                    data.EGEOD18494$time,".", 
                    substr(data.EGEOD18494$cell_line,1,1), ".",
                    data.EGEOD18494$rep)) %>% 
  write.csv(., file = "../data/exp.EGEOD18494.hif.csv", row.names = T)

```    


```{r}

exp.EGEOD18494.hif.pivot <- exp.EGEOD18494.hif %>%
  group_by(symbol) %>%
  pivot_longer(cols = starts_with("GSM"), names_to = "codes", values_to = "value") %>%
  merge(data.EGEOD18494, by = "codes") %>% 
  mutate(
    time = factor(time, levels = c("control", "4h" , "8h" , "12h"))
  )

```

# Comparing expression per cell line


```{r,  fig.height = 10, fig.width=15}

my_comparisons=list(c("control","4h"), c("control","8h"), c("control","12h"), c("4h","8h"), c("4h","12h"), c("8h","12h"))

ggplot(exp.EGEOD18494.hif.pivot[exp.EGEOD18494.hif.pivot$symbol == c('HIF1A'), ], aes(x = time, y = value, group = time, color = time)) +
    geom_jitter(alpha = 0.1, color = "tomato") +
    geom_boxplot() + facet_wrap(~ cell_line) + stat_compare_means(comparisons=my_comparisons)

ggplot(exp.EGEOD18494.hif.pivot[exp.EGEOD18494.hif.pivot$symbol == c('EPAS1'), ], aes(x = time, y = value, group = time, color = time)) +
    geom_jitter(alpha = 0.1, color = "tomato") +
    geom_boxplot() + facet_wrap(~ cell_line) + stat_compare_means(comparisons=my_comparisons)

```


```{r,  fig.height = 20, fig.width=15}


ggstatsplot::grouped_ggbetweenstats(
  data = exp.EGEOD18494.hif.pivot[exp.EGEOD18494.hif.pivot$symbol %in%  c('HIF1A', 'EPAS1', 'TP53', 'BNIP3'),],
  x = time,
  y = value,
  grouping.var = symbol, # grouping variable
  ggplot.component = list(ggplot2::scale_y_continuous(sec.axis = ggplot2::dup_axis())),
#  k = 4,
  title.prefix = "Gene",
  palette = "default_jama",
  package = "ggsci",
  plotgrid.args = list(nrow = 4),
  title.text = "Differences in gene expression for different time",
  messages = FALSE
)

```

# Another comparison visualization

```{r}

  ggplot(exp.EGEOD18494.hif.pivot[exp.EGEOD18494.hif.pivot$symbol == c('EPAS1', 'VEGFA'), ], aes(x = time, y = value, group = time, color = time)) +
    geom_jitter(alpha = 0.1, color = "tomato") +
    geom_boxplot() +
    facet_wrap(~ symbol)  +
    stat_compare_means(comparisons=my_comparisons)  +
    stat_compare_means(label.y = 25, vjust = 2)

  ggplot(exp.EGEOD18494.hif.pivot[exp.EGEOD18494.hif.pivot$symbol == c('CCNB1', 'BAD'), ], aes(x = time, y = value, group = time, color = time)) +
    geom_jitter(alpha = 0.1, color = "tomato") +
    geom_boxplot() +
    facet_wrap(~ symbol)  +
    stat_compare_means(comparisons=my_comparisons)  +
    stat_compare_means(label.y = 25, vjust = 2)

```


# Running DESeq2 Differential Expression


References:

https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/02_Preprocessing_Data.nb.html  

https://hbctraining.github.io/DGE_workshop_salmon/lessons/04_DGE_DESeq2_analysis.html

http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html 

https://bioconductor.github.io/BiocWorkshops/rna-seq-data-analysis-with-deseq2.html

http://bioconductor.org/packages/release/bioc/vignettes/Glimma/inst/doc/Glimma.pdf



<!-- ```{r} -->
<!-- save.image("pipeline_EGEOD18494.RData") -->
<!-- ``` -->

<!-- # Pre-processsing RNA-seq data -->

<!--     Based on:  -->

<!--      https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/02_Preprocessing_Data.nb.html  -->


<!-- # Convert counts to DESeqDataSet object -->


<!-- ```{r} -->
<!-- # create the design formula -->
<!--     design <- as.formula(~ condition) -->

<!-- # create the DESeqDataSet object -->
<!--     ddsObj <- DESeqDataSetFromMatrix(countData = exp.EGEOD18494, -->
<!--                                   colData = data.EGEOD18494, -->
<!--                                   design = design) -->

<!-- ``` -->


<!-- # Normalisation -->

<!-- ```{r} -->
<!-- # Apply normalisation to DDS object -->
<!--     ddsObj <- estimateSizeFactors(ddsObj) -->
<!-- ``` -->


<!-- ```{r} -->
<!--     ddsObj@colData$sizeFactor -->
<!-- ``` -->


<!-- # Interactive StripChart with Glimma -->


<!-- ```{r} -->
<!--     ddsObj <- DESeq(ddsObj) -->
<!--     res <- results(ddsObj) -->
<!--     res.shr <- DESeq2::lfcShrink(ddsObj, coef=2, res=res) -->

<!--     shrink.deseq <- as.data.frame(res.shr) %>% -->
<!--         rownames_to_column("probes") %>%  -->
<!--         left_join(anno.EGEOD18494, by="probes") %>%  -->
<!--         rename(logFC=log2FoldChange, FDR=padj) -->

<!-- ``` -->

<!-- ```{r} -->
<!--     summary(res) -->

<!-- ``` -->


<!-- ```{r} -->

<!--     hist(res.shr$pvalue , col = brewer.pal(3, name = "Set2")[1], -->
<!--          main = "Normoxia vs Hypoxia", xlab = "p-values") -->
<!-- ``` -->


<!-- ```{r} -->
<!-- # add a column with the names of only the top 30 genes -->
<!--     cutoff <- sort(res.shr$pvalue)[30] -->

<!-- # first remove the filtered genes (FDR=NA) and create a -log10(FDR) column -->
<!--     filtTab.deseq <- shrink.deseq %>% -->
<!--         filter(!is.na(FDR)) %>% -->
<!--         mutate(`-log10(FDR)` = -log10(FDR)) -->

<!--     filtTab.deseq <- filtTab.deseq %>%  -->
<!--         mutate(`-log10(FDR)`=pmin(`-log10(FDR)`, 51)) -->

<!-- ``` -->



<!-- ```{r} -->
<!--     hist(filtTab.deseq$pvalue, col = brewer.pal(3, name = "Set2")[1], -->
<!--          main = "Normoxia vs Hypoxia - Breast Cancer (Filtered)", xlab = "p-values") -->
<!-- ``` -->


<!-- ```{r} -->
<!-- # get the top genes -->
<!--     sigGenes <- as.data.frame(res.shr) %>%  -->
<!--         rownames_to_column("probes") %>%  -->
<!--         top_n(150, wt=-padj) %>%  -->
<!--         pull("probes") -->
<!-- # filter the data for the top 200 by padj in the LRT test -->
<!--     plotDat <- vst(ddsObj)[sigGenes,] %>%  -->
<!--         assay() -->
<!--     z.mat <- t(scale(t(plotDat), center=TRUE, scale=TRUE)) -->

<!-- # colour palette -->
<!--     myPalette <- c("red3", "ivory", "blue3") -->
<!--     myRamp = colorRamp2(c(-2, 0, 2), myPalette) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- # cluster the data and split the tree -->
<!--     hcDat <- hclust(dist(z.mat)) -->
<!--     cutGroups <- cutree(hcDat, h=4) -->
<!--     ha1 = HeatmapAnnotation(df = colData(ddsObj)[,c("cell_line", "cancer", "condition")]) -->
<!--     Heatmap(z.mat, name = "z-score", -->
<!--             col = myRamp,             -->
<!--             show_row_name = FALSE, -->
<!--             cluster_columns = FALSE, -->
<!--             split=cutGroups, -->
<!--             rect_gp = gpar(col = "darkgrey", lwd=0.5), -->
<!--             top_annotation = ha1, -->
<!--             row_names_gp = gpar(fontsize = 8), -->
<!--             column_names_gp = gpar(fontsize = 8), -->
<!--             column_names_rot = 45) -->

<!--     rm(hcDat,cutGroups, ha1 ) -->
<!-- ``` -->




<!-- ```{r} -->
<!--     de <- as.integer(filtTab.deseq$FDR <= 0.05) -->

<!--     normCounts <- log2(counts(ddsObj)) -->
<!--     filtCounts <- normCounts[filtTab.deseq$probes,] -->

<!--     glXYPlot( -->
<!--       x = filtTab.deseq$logFC, -->
<!--       y = -log10(filtTab.deseq$FDR), -->
<!--       xlab = "logFC", -->
<!--       ylab = "FDR", -->
<!--       main = "Normoxia v Hypoxia", -->
<!--       counts = filtCounts, -->
<!--       groups = data.EGEOD18494$condition, -->
<!--       status = de, -->
<!--       anno = filtTab.deseq[, c("probes", "symbol", "description")], -->
<!--       folder = "volcano_6647_DEseq2" -->
<!--     ) -->
<!-- ``` -->

<!-- # Selecting the HIF Genes  -->

<!-- ```{r} -->

<!--     filtTab.deseq_HIF <- filtTab.deseq[filtTab.deseq$symbol %in% hif.symbols, ] -->
<!-- ``` -->



<!-- ```{r} -->

<!--     ggplot(filtTab.deseq_HIF, aes(x = logFC, y=`-log10(FDR)`)) +  -->
<!--         geom_point(aes(colour=FDR < 0.05, shape = `-log10(FDR)` > 50), size=2) -->
<!-- ``` -->

<!-- # Showing a set of selected a Genes from Genecards -->

<!-- ```{r} -->
<!--     filtTab.deseq_HIF[, c("symbol", "logFC", "FDR", "description")] -->
<!-- ``` -->



<!-- ```{r} -->
<!--     group <- str_remove_all(data.EGEOD18494$condition, "[aeiou]") -->

<!--     de <- as.integer(filtTab.deseq_HIF$FDR <= 0.05) -->

<!--     normCounts <- log2(counts(ddsObj)) -->
<!--     filtCounts <- normCounts[filtTab.deseq_HIF$probes,] -->

<!--     glXYPlot( -->
<!--       x = filtTab.deseq_HIF$logFC, -->
<!--       y = -log10(filtTab.deseq_HIF$FDR), -->
<!--       xlab = "logFC", -->
<!--       ylab = "FDR", -->
<!--       main = "Normoxia v Hypoxia", -->
<!--       counts = filtCounts, -->
<!--       groups = group, -->
<!--       status = de, -->
<!--       anno = filtTab.deseq_HIF[, c("probes", "symbol", "description")], -->
<!--       folder = "volcano_6647_DEseq2_HIF" -->
<!--     ) -->
<!-- ``` -->


<!-- # Selecting the lower p-values  -->


<!-- ```{r} -->
<!--     nrow(subset(filtTab.deseq, pvalue < 0.001)) -->
<!-- ``` -->

<!-- ```{r}  -->
<!--     tail(subset(filtTab.deseq, pvalue < 0.001), 40)[, c("symbol", "logFC", "FDR", "pvalue", "description")] -->

<!-- ``` -->

<!-- # Selecting the lower  Adj. P-Values  -->


<!-- ```{r} -->
<!--     nrow(subset(filtTab.deseq, FDR < 0.001)) -->
<!-- ``` -->

<!-- ```{r}  -->
<!--     tail(subset(filtTab.deseq, FDR < 0.001), 30)[, c("symbol", "logFC", "FDR", "pvalue", "description")] -->

<!-- ``` -->



<!-- # Strip chart for gene expression -->

<!-- ```{r} -->
<!-- # Let's look at the most significantly differentially expressed gene -->
<!--     topgene <- filter(shrink.deseq, symbol=="HIF1A") -->
<!--     probes <- topgene$probes -->
<!--     plotCounts(ddsObj, gene = probes, intgroup = c("cancer", "condition"), -->
<!--                returnData = T) %>%  -->
<!--         ggplot(aes(x=condition, y=log2(count))) + -->
<!--           geom_point(aes(fill=condition), pch=21, size=2) + -->
<!--           facet_wrap(~cancer) + -->
<!--           expand_limits(y=0) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- # Let's look at the most significantly differentially expressed gene -->
<!--     topgene <- filter(shrink.deseq, symbol=="HIF1A") -->
<!--     probes <- topgene$probes -->
<!--     plotCounts(ddsObj, gene = probes, intgroup = c("cell_line", "condition"), -->
<!--                returnData = T) %>%  -->
<!--         ggplot(aes(x=condition, y=log2(count))) + -->
<!--           geom_point(aes(fill=condition), pch=21, size=2) + -->
<!--           facet_wrap(~cell_line) + -->
<!--           expand_limits(y=0) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- # Let's look at the most significantly differentially expressed gene -->
<!--     topgene <- filter(shrink.deseq, symbol=="TP53") -->
<!--     probes <- topgene$probes -->
<!--     plotCounts(ddsObj, gene = probes, intgroup = c("cancer", "condition"), -->
<!--                returnData = T) %>%  -->
<!--         ggplot(aes(x=condition, y=log2(count))) + -->
<!--           geom_point(aes(fill=condition), pch=21, size=2) + -->
<!--           facet_wrap(~cancer) + -->
<!--           expand_limits(y=0) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- # Let's look at the most significantly differentially expressed gene -->
<!--     topgene <- filter(shrink.deseq, symbol=="TP53") -->
<!--     probes <- topgene$probes -->
<!--     plotCounts(ddsObj, gene = probes, intgroup = c("cell_line", "condition"), -->
<!--                returnData = T) %>%  -->
<!--         ggplot(aes(x=condition, y=log2(count))) + -->
<!--           geom_point(aes(fill=condition), pch=21, size=2) + -->
<!--           facet_wrap(~cell_line) + -->
<!--           expand_limits(y=0) -->
<!-- ``` -->


<!-- # DEseq -->

<!-- ```{r} -->


<!--     shrink.deseq <- shrink.deseq %>%  -->
<!--        mutate(TopGeneLabel=ifelse(pvalue<=cutoff, symbol, "")) -->


<!--     ggplot(shrink.deseq, aes(x = log2(baseMean), y=logFC, label=TopGeneLabel)) +  -->
<!--         geom_point(aes(colour=FDR < 0.05), pch=20, size=0.5) + -->
<!--         labs(x="mean of normalised counts", y="log fold change") +  -->
<!--         geom_text_repel() -->
<!-- ``` -->


<!-- # Limma -->

<!-- ```{r} -->
<!--     cutoff <- sort(shrink.limma$pvalue)[20] -->

<!-- # first remove the filtered genes (FDR=NA) and create a -log10(FDR) column   -->
<!--     filtTab.limma <- shrink.limma %>%  -->
<!--         filter(!is.na(FDR)) %>%  -->
<!--         mutate(`-log10(FDR)` = -log10(FDR)) %>%  -->
<!--         mutate(TopGeneLabel=ifelse(pvalue<=cutoff, symbol, "")) -->

<!--     filtTab.limma <- filtTab.limma[!duplicated(filtTab.limma$probes),] -->
<!--     rownames(filtTab.limma) <- filtTab.limma$probes -->

<!--     ggplot(filtTab.limma, aes(x = logFC, y=`-log10(FDR)`, label=TopGeneLabel)) +  -->
<!--         geom_point(aes(colour=FDR < 0.05), size=2) + -->
<!--         geom_text_repel() -->
<!-- ``` -->


<!-- # DEseq -->


<!-- ```{r} -->
<!--     filtTab.deseq <- shrink.deseq %>%  -->
<!--         filter(!is.na(FDR)) %>%  -->
<!--         mutate(`-log10(FDR)` = -log10(FDR)) %>%  -->
<!--         mutate(TopGeneLabel=ifelse(pvalue<=cutoff, symbol, "")) -->

<!-- ``` -->


<!-- ```{r} -->

<!--     DE <- c("downregulated", "notDE", "upregulated")[as.factor(dt)] -->
<!--     anno <- as.data.frame(cbind(data.frame(probes=rownames(fit)), DE)) -->
<!--     anno <- as.data.frame(cbind(anno, filtTab.limma[anno$probes, c("symbol", "description")])) -->
<!--     rownames(anno) <- anno$probes     -->

<!--     glXYPlot( -->
<!--       x = filtTab.limma$logFC[anno$probes], -->
<!--       y = -log10(filtTab.limma$FDR[anno$probes]), -->
<!--       xlab = "logFC", -->
<!--       ylab = "FDR", -->
<!--       main = "Normoxia vs Hypoxia", -->
<!--       anno=anno,  -->
<!--       display.columns=c("probes", "symbol", "GeneName", "DE", "description"), -->
<!--       counts = filtCounts, -->
<!--       groups = data.EGEOD18494$condition, -->
<!--       status = de, -->
<!--       folder = "volcano_6647_limma" -->
<!--     ) -->

<!-- ``` -->



<!-- ```{r} -->

<!--     de <- as.integer(filtTab.limma$FDR[anno$probes] <= 0.05) -->

<!--     filtCounts <- exp.EGEOD18494[anno$probes,] -->

<!--     logFC <- filtTab.limma$logFC[filtTab.limma$probes %in% anno$probes] -->

<!--     cols <- c("yellow","blue","magenta") -->

<!--     sample.cols <- c("limegreen", "purple")[data.EGEOD18494$condition] -->

<!--     glXYPlot(x=fit$coef, y=fit$lod, xlab="logFC", ylab="logodds",status=dt, anno=anno, side.main="probes",counts=vm, groups=data.EGEOD18494$condition, sample.cols=sample.cols, -->
<!--              folder = "volcano_6647_limma_DE_condition" -->
<!--              ) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- deseq.df <- filtTab.deseq[, c("symbol", "logFC", "probes")] -->
<!-- limma.df  <- filtTab.limma[, c("symbol", "logFC", "probes")] -->

<!-- fdr <- p.adjust(fit$p.value[,  ncol(fit)], method = "BH") -->
<!-- limma.fit.df <- data.frame(probes = rownames(fit), logFC = fit$coefficients, fdr=fdr, baseMean=fit$Amean) -->
<!-- colnames(limma.fit.df) <- c("probes","logFC", "FDR", "baseMean") -->
<!-- limma.fit.df <- cbind(anno[limma.fit.df$probes, -1], limma.fit.df)  -->


<!-- save(deseq.df, limma.df, limma.fit.df, file = "DEA_Normoxia.vs.Hypoxia.rda") -->
<!-- rm(deseq.df, limma.df) -->
<!-- ``` -->


<!-- # Gene ontology (GO) based enrichment analysis - only DESeq2 -->

<!-- ```{r} -->
<!--     DE_genes <- subset(filtTab.deseq, FDR < 0.01)$probes -->

<!-- ``` -->



<!-- ```{r message=FALSE, warning=FALSE} -->
<!--     back_genes_idx <- genefilter::genefinder(exp.EGEOD18494,  -->
<!--                                             as.character(DE_genes),  -->
<!--                                             method = "manhattan", scale = "none") -->

<!-- ``` -->


<!-- ```{r} -->
<!--     back_genes_idx <- sapply(back_genes_idx, function(x)x$indices) -->

<!--     back_genes <- rownames(exp.EGEOD18494)[back_genes_idx] -->
<!--     back_genes <- setdiff(back_genes, DE_genes) -->


<!--     intersect(back_genes, DE_genes) -->
<!--     length(back_genes) -->
<!-- ``` -->


<!-- ```{r} -->
<!--     multidensity(list( -->
<!--             all = log(filtTab.deseq[,"baseMean"]), -->
<!--             fore = log(filtTab.deseq[filtTab.deseq$probes %in% DE_genes , "baseMean"]), -->
<!--             back = log(filtTab.deseq[filtTab.deseq$probes %in% back_genes, "baseMean"])), -->
<!--             col = c("#e46981", "#ae7ee2", "#a7ad4a"), -->
<!--          xlab = "mean expression", -->
<!--        main = "DE genes for Hypoxia background-matching") -->
<!-- ``` -->


<!--     When comparing the “background gene” curve to the “foreground gene” curve, we see a similar curve shape, indicating a sensible background matching (Figure above). Note that the right-shift of the “foreground-gene” curve in comparison to the “background-gene” curve indicates that DE-genes are generally very highly expressed, so that it wasn’t possible to find background-genes with exactly equal overall expression distribution. -->

<!--     The “all gene” curve has the leftmost curve maximum with a bimodality; this can be explained by a high number of lowly expressed genes in all samples and shows that a foreground matching is sensible in order to avoid biases.  -->


<!-- # Running topGO - Only DEseq  -->

<!-- ```{r} -->
<!--     gene_IDs <- filtTab.deseq$probes -->
<!--     in_universe <- gene_IDs %in% c(DE_genes, back_genes) -->
<!--     in_selection <- gene_IDs %in% DE_genes  -->

<!--     all_genes <- in_selection[in_universe] -->
<!--     all_genes <- factor(in_selection[in_universe]) -->
<!--     names(all_genes) <- gene_IDs[in_universe]  -->

<!-- ``` -->


<!-- ```{r} -->
<!--     top_GO_data <- new("topGOdata", ontology = "BP", allGenes = all_genes, -->
<!--      nodeSize = 10, annot=annFUN.org, mapping="org.Hs.eg.db", ID="ensgene") -->

<!-- ``` -->

<!-- ```{r} -->
<!--     result_top_GO_elim <-  -->
<!--       runTest(top_GO_data, algorithm = "elim", statistic = "Fisher") -->

<!--     result_top_GO_classic <-  -->
<!--       runTest(top_GO_data, algorithm = "classic", statistic = "Fisher") -->
<!-- ``` -->


<!-- ```{r} -->

<!--     library(Biobase) -->
<!--     library(org.Hs.eg.db) -->
<!--     #library(org.Hs.egensgene) -->
<!--     res_top_GO <- GenTable(top_GO_data, Fisher.elim = result_top_GO_elim, -->
<!--             Fisher.classic = result_top_GO_classic, -->
<!--             orderBy = "Fisher.elim" , topNodes = 100) -->

<!-- # genes_top_GO <- printGenes(top_GO_data, whichTerms = res_top_GO$GO.ID, -->
<!-- #    chip = "org.Hs.eg.db",  geneCutOff = 1000) -->
<!-- #  -->
<!-- # res_top_GO$sig_genes <- sapply(genes_top_GO, function(x){ -->
<!-- #                str_c(paste0(x[x$'raw p-value' == 2, "symbol.id"],";"), -->
<!-- #                      collapse = "") -->
<!-- #    }) -->

<!--     head(res_top_GO[,1:8], 20) -->
<!-- ``` -->




<!-- ```{r} -->
<!--     ensgene_ids <- mapIds(org.Hs.eg.db, -->
<!--           keys = filtTab.deseq$probes, -->
<!--           keytype = "ensgene", -->
<!--           column = c("ENTREZID")) -->

<!--     ensgene_ids <- ensgene_ids[!is.na(ensgene_ids)] -->
<!-- ``` -->


<!-- ```{r} -->
<!--     reactome_enrich <- enrichPathway(gene = ensgene_ids[DE_genes], -->
<!--                                     universe = ensgene_ids[c(DE_genes, -->
<!--                                                             back_genes)], -->
<!--                                     organism = "human", -->
<!--                                     pvalueCutoff = 0.05, -->
<!--                                     qvalueCutoff = 0.9, -->
<!--                                     readable = TRUE) -->

<!--     reactome_enrich@result$description <- paste0(str_sub( -->
<!--                                         reactome_enrich@result$description, 1, 20), -->
<!--                                         "...") -->

<!--     head(as.data.frame(reactome_enrich))[1:6] -->
<!--     #save.image("maEndToEnd_6647_pdf.RData") -->
<!-- ``` -->

<!-- ```{r} -->
<!--     barplot(reactome_enrich) -->

<!-- ``` -->


<!-- ```{r} -->
<!--     emapplot(reactome_enrich, showCategory = 10) -->
<!-- ``` -->



<!-- # Gene ontology (GO) based enrichment analysis with deSeq and Limma -->

<!-- ```{r} -->
<!--     DE_genes <- subset(filtTab.limma[filtTab.deseq$probes, ], FDR < 0.01)$probes -->

<!-- ``` -->



<!-- ```{r message=FALSE, warning=FALSE} -->
<!--     back_genes_idx <- genefilter::genefinder(exp.EGEOD18494,  -->
<!--                                             as.character(DE_genes),  -->
<!--                                             method = "manhattan", scale = "none") -->

<!-- ``` -->


<!-- ```{r} -->
<!--     back_genes_idx <- sapply(back_genes_idx, function(x)x$indices) -->

<!--     back_genes <- rownames(exp.EGEOD18494)[back_genes_idx] -->
<!--     back_genes <- setdiff(back_genes, DE_genes) -->


<!--     intersect(back_genes, DE_genes) -->
<!--     length(back_genes) -->
<!-- ``` -->


<!-- ```{r} -->
<!--     multidensity(list( -->
<!--             all = log(filtTab.deseq[,"baseMean"]), -->
<!--             fore = log(filtTab.deseq[filtTab.deseq$probes %in% DE_genes , "baseMean"]), -->
<!--             back = log(filtTab.deseq[filtTab.deseq$probes %in% back_genes, "baseMean"])), -->
<!--             col = c("#e46981", "#ae7ee2", "#a7ad4a"), -->
<!--          xlab = "mean expression", -->
<!--        main = "DE genes for Hypoxia background-matching") -->
<!-- ``` -->


<!--     When comparing the “background gene” curve to the “foreground gene” curve, we see a similar curve shape, indicating a sensible background matching (Figure above). Note that the right-shift of the “foreground-gene” curve in comparison to the “background-gene” curve indicates that DE-genes are generally very highly expressed, so that it wasn’t possible to find background-genes with exactly equal overall expression distribution. -->

<!--     The “all gene” curve has the leftmost curve maximum with a bimodality; this can be explained by a high number of lowly expressed genes in all samples and shows that a foreground matching is sensible in order to avoid biases.  -->


<!-- # Running topGO - Intersection DEseq and Limma -->

<!-- ```{r} -->
<!--     gene_IDs <- filtTab.limma$probes[filtTab.limma$probes %in% filtTab.deseq$probes] -->
<!--     in_universe <- gene_IDs %in% c(DE_genes, back_genes) -->
<!--     in_selection <- gene_IDs %in% DE_genes  -->

<!--     all_genes <- in_selection[in_universe] -->
<!--     all_genes <- factor(in_selection[in_universe]) -->
<!--     names(all_genes) <- gene_IDs[in_universe]  -->

<!-- ``` -->


<!-- ```{r} -->
<!--     top_GO_data <- new("topGOdata", ontology = "BP", allGenes = all_genes, -->
<!--      nodeSize = 10, annot=annFUN.org, mapping="org.Hs.eg.db", ID="ensgene") -->

<!-- ``` -->

<!-- ```{r} -->
<!--     result_top_GO_elim <-  -->
<!--       runTest(top_GO_data, algorithm = "elim", statistic = "Fisher") -->
<!--     result_top_GO_classic <-  -->
<!--       runTest(top_GO_data, algorithm = "classic", statistic = "Fisher") -->
<!-- ``` -->


<!-- ```{r} -->

<!--     library(Biobase) -->
<!--     library(org.Hs.eg.db) -->
<!--     #library(org.Hs.egensgene) -->
<!--     res_top_GO <- GenTable(top_GO_data, Fisher.elim = result_top_GO_elim, -->
<!--             Fisher.classic = result_top_GO_classic, -->
<!--             orderBy = "Fisher.elim" , topNodes = 100) -->

<!-- # genes_top_GO <- printGenes(top_GO_data, whichTerms = res_top_GO$GO.ID, -->
<!-- #    chip = "org.Hs.eg.db",  geneCutOff = 1000) -->
<!-- #  -->
<!-- # res_top_GO$sig_genes <- sapply(genes_top_GO, function(x){ -->
<!-- #                str_c(paste0(x[x$'raw p-value' == 2, "symbol.id"],";"), -->
<!-- #                      collapse = "") -->
<!-- #    }) -->

<!--     head(res_top_GO[,1:8], 20) -->
<!-- ``` -->




<!-- ```{r} -->
<!--     ensgene_ids <- mapIds(org.Hs.eg.db, -->
<!--           keys = filtTab.limma$probes, -->
<!--           keytype = "ensgene", -->
<!--           column = c("ENTREZID")) -->

<!--     ensgene_ids <- ensgene_ids[!is.na(ensgene_ids)] -->
<!-- ``` -->


<!-- ```{r} -->
<!--     reactome_enrich <- enrichPathway(gene = ensgene_ids[DE_genes], -->
<!--                                     universe = ensgene_ids[c(DE_genes, -->
<!--                                                             back_genes)], -->
<!--                                     organism = "human", -->
<!--                                     pvalueCutoff = 0.05, -->
<!--                                     qvalueCutoff = 0.9, -->
<!--                                     readable = TRUE) -->

<!--     reactome_enrich@result$description <- paste0(str_sub( -->
<!--                                         reactome_enrich@result$description, 1, 20), -->
<!--                                         "...") -->

<!--     head(as.data.frame(reactome_enrich))[1:6] -->
<!--     #save.image("maEndToEnd_6647_pdf.RData") -->
<!-- ``` -->

<!-- ```{r} -->
<!--     barplot(reactome_enrich) -->

<!-- ``` -->


<!-- ```{r} -->
<!--     emapplot(reactome_enrich, showCategory = 10) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- save.image("maEndToEnd_6647.RData") -->
<!-- ``` -->


